sql/core/src/main/scala/org/apache/spark/sql/execution/RemoveRedundantProjects.scala

```diff
@@ -61,15 +63,24 @@ object RemoveRedundantProjects extends Rule[SparkPlan] {
         val keepOrdering = a.aggregateExpressions
           .exists(ae => ae.mode.equals(Final) || ae.mode.equals(PartialMerge))
         a.mapChildren(removeProject(_, keepOrdering))
-      // GenerateExec requires column ordering since it binds input rows directly with its
-      // requiredChildOutput without using child's output schema.
-      case g: GenerateExec => g.mapChildren(removeProject(_, true))
-      // JoinExec ordering requirement will inherit from its parent. If there is no ProjectExec in
-      // its ancestors, JoinExec should require output columns to be ordered.
-      case o => o.mapChildren(removeProject(_, requireOrdering))
+      case p if canPassThrough(p) => p.mapChildren(removeProject(_, requireOrdering))
```

Comment:
Why didn't you simply add `case u: UnionExec => u.mapChildren(removeProject(_, true))`?

Reply:
Good question. This is because there might be other nodes that require column ordering. An earlier PR also fixed a similar issue for GenerateExec. It is safer to use a whitelist in order to prevent similar issues. Most existing cases are covered (filter, aggregate and join). We can always add more nodes to the whitelist if redundant projects are discovered.

Reply:
If so, I think you'd be better to clearly describe what this PR proposes. The current title and description looks a bit confusing. It seems this PR affects the optimization behavior for all the physical nodes other than filter, expand, aggregate and join.